packages/core/src/utils/mergeAbis.ts

```diff
@@ -0,0 +1,50 @@
+import type { Abi } from "abitype";
+import type { AbiItem } from "viem";
+import { formatAbiItem } from "viem/utils";
+
+export type MergeAbi<
+  TBase extends Abi,
+  TInsert extends Abi,
+> = TInsert extends readonly [
+  infer First extends AbiItem,
+  ...infer Rest extends Abi,
+]
+  ? Extract<TBase[number], First> extends never
+    ? MergeAbi<readonly [...TBase, First], Rest>
+    : MergeAbi<TBase, Rest>
+  : TBase;
+
+type MergeAbis<
+  TMerged extends Abi,
+  TImpls extends readonly Abi[],
+> = TImpls extends readonly [
+  infer First extends Abi,
+  ...infer Rest extends readonly Abi[],
+]
+  ? MergeAbis<MergeAbi<TMerged, First>, Rest>
+  : TMerged;
+
+const isAbiEqual = (a: AbiItem, b: AbiItem): boolean =>
+  formatAbiItem(a) === formatAbiItem(b);
+
+/**
+ * Build a single Abi from a proxy and its implementations
+ */
+export const mergeAbis = <
+  const TProxy extends Abi,
+  const TImpl extends readonly Abi[],
```

Comment:
Nit: This function could be useful in cases other than proxy/impl, so if possilbe would probably just have one generic parameter `TAbis extends readonly Abi[]` ?